lib/ecto/repo/schema.ex

```diff
@@ -76,6 +81,40 @@ defmodule Ecto.Repo.Schema do
     end
   end
 
+  defp extract_header_and_fields(query = %Ecto.Query{}, _schema, _dumper, _autogen_id, _placeholder_map, adapter) do
+    case query.select do
+      nil ->
+        raise ArgumentError, message: """
+Cannot generate a fields list for INSERT INTO from the given source query:
+#{inspect query}
+
+Please add a select clause that selects into a map, like this:
+
+from x in Source,
+  ...,
+  select: %{
+    field_a: x.bar,
+    field_b: x.foo
+  }
+        """
+
+      %Ecto.Query.SelectExpr{expr: {op, _ctx, args}} when op in [:%{}, :"[]"] ->
+        header =
+          Keyword.keys(args)
+          |> Enum.reduce(%{}, &Map.put(&2, &1, true))
+
+        query = Map.update!(query, :select, fn select ->
+          Map.update!(select, :expr, fn {op, ctx, args} ->
+            args = Enum.sort_by(args, &elem(&1, 0))
```

Comment:
The sort is necessary since the `VALUES` clause in SQL is fed by a map, where the keys are alphabetically sorted.